### ISSUE TITLE
`Runtime::shutdown_timeout` unnecessarily blocks until the timeout expires

### ISSUE THREAD

**Issue by dthul** [tokio-rs/tokio#2314]
## Version
tokio v0.2.13

## Platform
Darwin 19.3.0 Darwin Kernel Version 19.3.0: Thu Jan  9 20:58:23 PST 2020; root:xnu-6153.81.5~1/RELEASE_X86_64 x86_64

## Description
I want to gracefully shutdown my program. In order to do so, I first drop my long-lived "root futures" (like a warp server) and then call `Runtime::shutdown_timeout` to wait for short-lived tasks that were spawned onto the runtime to finish (such as email sending tasks).

I expected `Runtime::shutdown_timeout` to return as soon as all spawned tasks finish, but this does not seem to be the case. Instead, it always seems to block until the specified timeout expires.

I created a small self-contained example illustrating the problem:

Dependencies:
```toml
tokio = { version = "0.2", default-features = false, features = ["io-driver", "rt-threaded", "time"] }
```

Code:
```rust
fn main() {
    let mut async_runtime = tokio::runtime::Builder::new()
        .enable_io()
        .enable_time()
        .threaded_scheduler()
        .build()
        .expect("Could not create tokio runtime");
    async_runtime.block_on(async {
        tokio::spawn(async {
            let mut i = 5;
            while i > 0 {
                println!("{}", i);
                tokio::time::delay_for(tokio::time::Duration::from_secs(1)).await;
                i -= 1;
            }
            println!("Async task done")
        });
    });
    println!("Shutting down the runtime");
    async_runtime.shutdown_timeout(tokio::time::Duration::from_secs(10));
    println!("Runtime shut down");
}
```

When running this snippet you can observe that even after the spawned task has finished after 5 seconds, the `shutdown_timeout` call still waits for the full 10 seconds before it returns.

**Comment by hlavaatch** [comment]
It is getting stuck in `tokio::runtime::Enter::block_on_timeout`, parked, not waking up because lack of notifications.

**Comment by pawanjay176** [comment]
+1 on this..facing the same issue in `tokio v0.2.21`

**Comment by drewkett** [comment]
I just ran into this same error. Here's an even simpler example that still behaves the same way in case its helpful
```rust
fn main() {
    let mut rt = tokio::runtime::Runtime::new().unwrap();
    rt.block_on(async {});
    let t = std::time::Instant::now();
    rt.shutdown_timeout(std::time::Duration::from_secs(2));
    println!("Took {} ms", (std::time::Instant::now() - t).as_millis())
}
```

**Comment by drewkett** [comment]
Also, it only occurs with `rt-threaded` enabled

**Comment by hlavaatch** [comment]
Did a bit more digging, instrumenting tokio.
It seems shutdown logic is being subverted by thread parking mechanism in `tokio::runtime::thread_pool::worker::Context::run()` .

Worker threads do not respond to the shutdown notification because they are stuck inside task.run() call in  `tokio::runtime::blocking::pool::Inner::run(&self)` that does not return - the thread gets parked instead. Did not see single worker thread return from that.
None of the shutdown, IDLE state logic and thread state counting in Inner::run() is being run as a result.
Hence, worker threads never drop the shutdown_tx handles, so shutdown_timeout() waits for maximum time.

We probably need to somehow get the threads to unpark on shutdown?

**Comment by ajfrantz** [comment]
I believe this issue was fixed in #2649. The test cases listed above seem to work as expected for me on v1.0.0.
